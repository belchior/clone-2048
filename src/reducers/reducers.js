import { loadState, saveState, } from '../localState';
import { last, pipe, reverse, tail, raffle, } from '../helpers/list';
import {
  LOAD,
  MOVE_ERROR,
  MOVIMENT,
  PLAYER_LOSE,
  PLAYER_WON,
  PLAYING,
  RESTART,
  ROLLBACK,
  SAVE,
  START,
  TOGGLE_HARDMODE,
  WELCOME
} from './actions/types';


export const initialState = {
  bestScore: 0,
  hardMode: false,
  history: [],
  maxBlock: 2048,
  moveError: false,
  rollback: 2,
  score: 0,
  status: PLAYING,
  wall: [],
  welcomeWall: [ 2, 4, 8, 16, 4096, 0, 0, 32, 2048, 0, 0, 64, 1024, 512, 256, 128, ],
};

const initialize = () => ({
  ...initialState,
  status: WELCOME,
});

export const reducer = (state = initialize(), action) => {
  switch (action.type) {
    case LOAD: return load(state);
    case MOVIMENT: return moviment(state, action.payload);
    case MOVE_ERROR: return moveError(state, action.payload);
    case PLAYER_LOSE: return playerLose(state);
    case PLAYER_WON: return playerWon(state);
    case RESTART: return restart(state);
    case ROLLBACK: return rollback(state);
    case SAVE: return save(state);
    case START: return start(state);
    case TOGGLE_HARDMODE: return toggleHardMode(state);
    default: return state;
  }
};


const load = state => ({
  ...loadState(),
  ...state,
});

const moviment = (state, newState) => {
  return pipe(
    setBestScore,
    addHistory(state),
  )(newState);
};

const addHistory = state => newState => {
  const lastWall = last(state.history);
  return {
    ...newState,
    history: lastWall ? [ lastWall, state.wall ] : [ state.wall ],
  };
};

const save = state => {
  saveState(state);
  return state;
};

const start = state => ({
  ...state,
  history: [],
  wall: raffle(raffle(Array(16).fill(0))),
  status: PLAYING,
});

const setBestScore = state => ({
  ...state,
  bestScore: state.score > state.bestScore
    ? state.score
    : state.bestScore,
});


const moveError = (state, error) => ({
  ...state,
  moveError: error
});


const playerLose = state => ({
  ...state,
  status: PLAYER_LOSE,
});


const playerWon = state => ({
  ...state,
  status: PLAYER_WON,
});

const restart = state => ({
  ...initialState,
  bestScore: state.bestScore,
  history: [],
  wall: raffle(raffle(Array(16).fill(0))),
  status: PLAYING,
});

const rollback = state => {
  if (state.history.length <= 0 || state.rollback <= 0 ) return state;
  return {
    ...state,
    rollback: Math.max(state.rollback - 1, 0),
    wall: last(state.history),
    history: tail(reverse(state.history)),
  };
};

const toggleHardMode = state => ({
  ...state,
  hardMode: !state.hardMode
});
