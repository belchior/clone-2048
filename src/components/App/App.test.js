import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import { App } from './index';
import { initialState } from '../../reducers/reducers';
const renderer = new ShallowRenderer();


it('App should render Welcome without crashing', () => {
  const dispatch = jest.fn(action => action);
  const state = {
    ...initialState,
    status: 'WELCOME',
  };
  const tree = renderer.render(<App state={state} dispatch={dispatch} />);

  expect(tree).toMatchSnapshot();
});

it('App should render Wall without crashing', () => {
  const dispatch = jest.fn(action => action);
  const state = {
    ...initialState,
    status: 'PLAYING',
  };
  const tree = renderer.render(<App state={state} dispatch={dispatch} />);

  expect(tree).toMatchSnapshot();
});

it('App should render Wall without crashing when status id not defined', () => {
  const dispatch = jest.fn(action => action);
  const state = {
    ...initialState,
    status: undefined,
  };
  const tree = renderer.render(<App state={state} dispatch={dispatch} />);

  expect(tree).toMatchSnapshot();
});

it('App should render PlayerLose without crashing', () => {
  const dispatch = jest.fn(action => action);
  const state = {
    ...initialState,
    status: 'PLAYER_LOSE',
  };
  const tree = renderer.render(<App state={state} dispatch={dispatch} />);

  expect(tree).toMatchSnapshot();
});
