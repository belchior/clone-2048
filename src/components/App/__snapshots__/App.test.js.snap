// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render PlayerLose without crashing 1`] = `
<div
  className="App no-select"
>
  <Connect(Sidebar) />
  <main
    className="main"
  >
    <Connect(PlayerLose) />
  </main>
</div>
`;

exports[`App should render Wall without crashing 1`] = `
<Keyboard
  shortcuts={
    Array [
      Object {
        "action": [Function],
        "shortcut": "arrow-down",
      },
      Object {
        "action": [Function],
        "shortcut": "arrow-left",
      },
      Object {
        "action": [Function],
        "shortcut": "arrow-right",
      },
      Object {
        "action": [Function],
        "shortcut": "arrow-up",
      },
    ]
  }
>
  <div
    className="App no-select"
  >
    <Connect(Sidebar) />
    <main
      className="main"
    >
      <Connect(Wall) />
    </main>
  </div>
</Keyboard>
`;

exports[`App should render Wall without crashing when status id not defined 1`] = `
<Keyboard
  shortcuts={
    Array [
      Object {
        "action": [Function],
        "shortcut": "arrow-down",
      },
      Object {
        "action": [Function],
        "shortcut": "arrow-left",
      },
      Object {
        "action": [Function],
        "shortcut": "arrow-right",
      },
      Object {
        "action": [Function],
        "shortcut": "arrow-up",
      },
    ]
  }
>
  <div
    className="App no-select"
  >
    <Connect(Sidebar) />
    <main
      className="main"
    >
      <Connect(Wall) />
    </main>
  </div>
</Keyboard>
`;

exports[`App should render Welcome without crashing 1`] = `
<div
  className="App no-select"
>
  <Connect(Sidebar) />
  <main
    className="main"
  >
    <Connect(Welcome) />
  </main>
</div>
`;
